flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManager.java

```diff
@@ -0,0 +1,381 @@
+		return workerNodeMap.get(resourceID);
+	}
+
+	@Override
+	public boolean stopWorker(final KubernetesWorkerNode worker) {
+		Preconditions.checkNotNull(this.kubeClient);
+		LOG.info("Stopping Worker {}.", worker.getResourceID().toString());
+		try {
+			this.kubeClient.stopPod(worker.getResourceID().toString());
+		} catch (Exception e) {
+			this.kubeClient.handleException(e);
+			return false;
+		}
+		workerNodeMap.remove(worker.getResourceID());
+		return true;
+	}
+
+	@Override
+	public void onAdded(List<KubernetesPod> pods) {
+		runAsync(() -> {
+			for (KubernetesPod pod : pods) {
+				log.info("Received new TaskExecutor pod: {} - Remaining pending pod requests: {}",
+					pod.getName(), numPendingPodRequests);
+
+				if (numPendingPodRequests > 0) {
+					numPendingPodRequests--;
+					final KubernetesWorkerNode worker = new KubernetesWorkerNode(new ResourceID(pod.getName()));
+					workerNodeMap.putIfAbsent(worker.getResourceID(), worker);
+				}
+			}
+		});
+	}
+
+	@Override
+	public void onModified(List<KubernetesPod> pods) {
+		runAsync(() -> pods.forEach(this::removePodIfTerminated));
+	}
+
+	@Override
+	public void onDeleted(List<KubernetesPod> pods) {
+		runAsync(() -> pods.forEach(this::removePodIfTerminated));
+	}
+
+	@Override
+	public void onError(List<KubernetesPod> pods) {
+		runAsync(() -> pods.forEach(this::removePodIfTerminated));
+	}
+
+	@VisibleForTesting
+	ConcurrentMap<ResourceID, KubernetesWorkerNode> getWorkerNodeMap() {
+		return workerNodeMap;
+	}
+
+	private void getWorkerNodesFromPreviousAttempts() throws Exception {
+		final List<KubernetesPod> podList = kubeClient.getPodsWithLabels(getTaskManagerLabels());
+		for (KubernetesPod pod : podList) {
+			final KubernetesWorkerNode worker = new KubernetesWorkerNode(new ResourceID(pod.getName()));
+			workerNodeMap.put(worker.getResourceID(), worker);
+			final long attempt = worker.getAttempt();
+			if (attempt > currentMaxAttemptId.get()) {
+				currentMaxAttemptId.set(attempt);
+			}
+		}
+
+		log.info("Recovered {} pods from previous attempts, current attempt id is {}.",
+			workerNodeMap.size(),
+			currentMaxAttemptId.addAndGet(1));
+	}
+
+	private void requestKubernetesPod() {
+		Preconditions.checkNotNull(this.kubeClient);
+
+		numPendingPodRequests++;
+
+		log.info("Requesting new TaskExecutor pod with <{},{}>. Number pending requests {}.",
+			defaultTaskManagerMemoryMB,
+			defaultCpus,
+			numPendingPodRequests);
+
+		final String podName = String.format(
+			TASK_MANAGER_POD_FORMAT,
+			clusterId,
+			currentMaxAttemptId.get(),
+			currentMaxPodId.addAndGet(1));
+
+		try {
+			final HashMap<String, String> env = new HashMap<>();
+			env.put(Constants.ENV_FLINK_POD_NAME, podName);
+
+			final TaskManagerPodParameter parameter = new TaskManagerPodParameter(
+				podName,
+				getTaskManagerStartCommand(podName),
+				defaultTaskManagerMemoryMB,
+				defaultCpus,
+				env);
+
+			this.kubeClient.createTaskManagerPod(parameter);
```

Comment:
The clauses before this can be moved out of `try`. Actually the interface of `createTaskManagerPod` also does not throw any exceptions, so i am not quite sure whether it might throw some implicit exception in practice.

Reply:
It will not throw any exceptions here. I will remove the `try..catch`